File: Problem53_ReorderList.java
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
class Solution {
    public void reorderList(ListNode head) {
        if(head == null)
            return;
        ListNode slow = head;
        ListNode fast = head;
        while(fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        ListNode prev = null;
        while(slow != null) {
            ListNode temp = slow.next;
            slow.next = prev;
            prev = slow;
            slow = temp;
        }
        ListNode curr = head;
        while(prev.next != null) {
            ListNode temp = curr.next;
            curr.next = prev;
            curr = temp;
            
            temp = prev.next;
            prev.next = curr;
            prev = temp;
        }
    }
}

File: Problem-53 Reorder List.py
# 143. Reorder List
# https://leetcode.com/problems/reorder-list/

# Time Complexity: O(n)
# Space Complexiety: O(1)

# Logic: We need to reverse the second half of the LL. 
# Then using both the heads, insert the nodes.

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverse(self,head):
        prev, cur = None, head
        
        while cur:
            temp = cur.next
            cur.next = prev
            prev = cur
            cur = temp
        
        return prev
    
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        # find mid of the LL
        slow, fast = head, head
        
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        
        # Reverse elements after mid
        head2 = self.reverse(slow.next)
        slow.next = None
            
        ptr = head
        # Insert elements
        while head2:
            temp = head.next
            head.next = head2
            head = temp
            temp = head2.next
            head2.next = head
            head2 = temp
        
        return ptr